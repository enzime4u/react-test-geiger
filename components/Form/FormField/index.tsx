import React from "react";
import FormFieldStructure from "./FormFieldStructure";
import { FormFieldProps } from "./FormFieldTypes";

const FormField = (props:FormFieldProps) => {
  return <FormFieldStructure {...props} />;
};

export default FormField;
