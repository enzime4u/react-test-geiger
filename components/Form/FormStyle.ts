import styled from "styled-components";

export const FormWrapper = styled.form`
  display: flex;
  flex-direction: column;
  border: 1px solid #ccc;
  padding: 2rem;
  border-radius: 0.5rem;
  gap: 1rem;
  width: 400px;

  @media (max-width: 399px) {
    width: 350px;
  }
`;

export const FormSubmitButton = styled.button`
  padding: 0.6rem;
  background-color: #a8c0d0;
  border: none;
  border-radius: 0.3rem;
  margin-top: 1rem;
`;
