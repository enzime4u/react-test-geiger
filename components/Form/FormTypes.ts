type FormFiledsTypes = "text" | "select" | "checkbox";

export type FormProps = {
  label: string;
  type: string;
  name: string;
  placeholder?: string;
  options?: string[];
  data: {
    type: FormFiledsTypes;
    initialValue: any;
    name: string;
    label: string;
    options: { label: string; value: string }[];
  }[];
};
