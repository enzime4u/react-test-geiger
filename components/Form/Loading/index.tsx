import React from "react";
import LoadingStructure from "./LoadingStructure";
import "./LoadingStyles.ts";

const Loading: React.FC = () => {
  return <LoadingStructure />;
};

export default Loading;
