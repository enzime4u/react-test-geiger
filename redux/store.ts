import { configureStore } from '@reduxjs/toolkit';
import formSlice from './formSlice';
import alertSlice from './alertSlice';

const store = configureStore({
  reducer: {
    newsLetter: formSlice,
    alert: alertSlice,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
