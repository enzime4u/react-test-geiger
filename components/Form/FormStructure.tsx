import React from "react";
import FormField from "./FormField";
import { FormWrapper, FormSubmitButton } from "./FormStyle";
import { useDispatch, useSelector } from "react-redux";
import { resetForm, submitForm, setErrorFields } from "@/redux/formSlice";
import { showAlert, hideAlert } from "@/redux/alertSlice";
import { EMAIL_REGEX } from "@/utils/validations";
import { RootState } from "@/redux/store";

const formData = [
  {
    type: "email",
    placeholder: "Add your email",
    name: "email",
    label: "Email",
  },
  {
    type: "select",
    label: "Please select a category",
    placeholder: "Choose",
    name: "select",
    options: [
      { value: "-", label: "Select an option" },
      { value: "Shoes", label: "Shoes" },
      { value: "Heels", label: "Heels" },
      { value: "Sandals", label: "Sandals" },
      { value: "Handbags", label: "Handbags" },
    ],
  },
  {
    type: "checkbox",
    name: "checkbox",
    checked: false,
    label: "Tick if you agree with terms and conditions",
  },
];

const FormStructure = () => {
  const dispatch = useDispatch();
  const form = useSelector((state: RootState) => state.newsLetter);
  const isValidForm = () => {
    return EMAIL_REGEX.test(form.email) && form.agreedToTerms;
  };

  const handleFormSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();

    if (isValidForm()) {
      dispatch(submitForm(form));
      dispatch(showAlert("Form submitted successfully"));
      dispatch(resetForm());
      setTimeout(() => dispatch(hideAlert()), 5000);
    } else {
      dispatch(
        setErrorFields({
          emailErr: !EMAIL_REGEX.test(form.email),
          checkboxErr: !form.agreedToTerms,
        }),
      );
      dispatch(showAlert("Error submitting form, please check fields"));
      setTimeout(() => dispatch(hideAlert()), 2500);
    }
  };

  return (
    <FormWrapper>
      {formData.map((field) => (
        <FormField
          key={field.name}
          label={field.label}
          type={field.type}
          name={field.name}
          placeholder={field?.placeholder}
          options={field?.options}
        />
      ))}
      <FormSubmitButton onClick={handleFormSubmit} type="submit">
        Submit
      </FormSubmitButton>
    </FormWrapper>
  );
};

export default FormStructure;
