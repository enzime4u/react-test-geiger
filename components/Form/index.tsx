import React from 'react';
import FormStructure from './FormStructure';

const Form = () => {
  return (
    <div>
      <FormStructure />
    </div>
  );
};

export default Form;
