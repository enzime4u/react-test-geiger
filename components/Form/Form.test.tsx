import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import store from "../../redux/store";
import { ThemeProvider } from "styled-components";
import { theme } from "@/configs";
import Form from ".";

const withProviders = (children: React.JSX.Element) => (
  <ThemeProvider theme={theme}>
    <Provider store={store}>{children}</Provider>
  </ThemeProvider>
);

it("it should match the snapshot and render", async () => {
  const { container } = render(withProviders(<Form />));

  expect(container).toMatchSnapshot();
  expect(container).toBeInTheDocument();

  const emailInput = await screen.findByPlaceholderText("Add your email");
  const selectInput = await screen.findByPlaceholderText("Choose");
  const checkboxInput = screen.getByRole("checkbox");

  fireEvent.change(emailInput, { target: { value: "test@test.com" } });
  fireEvent.change(selectInput, { target: { value: "Shoes" } });
  fireEvent.click(checkboxInput!);

  expect(emailInput).toHaveValue("test@test.com");
  expect(selectInput).toHaveValue("Shoes");
  expect(checkboxInput).toBeChecked();
});
