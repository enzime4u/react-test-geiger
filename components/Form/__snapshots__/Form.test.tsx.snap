// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match the snapshot and render 1`] = `
<div>
  <div>
    <form
      class="sc-4142dcf-0 bukgte"
    >
      <label
        class="sc-c25ec0ba-0 AOPnj"
        for=":r0:"
      >
        Email
      </label>
      <input
        class="sc-c25ec0ba-1 kunqTV error"
        id=":r0:"
        label="Email"
        name="email"
        placeholder="Add your email"
        type="email"
        value=""
      />
      <label
        class="sc-c25ec0ba-0 AOPnj"
        for=":r1:"
      >
        Please select a category
      </label>
       
      <select
        class="sc-c25ec0ba-2 iQdcvS"
        id=":r1:"
        label="Please select a category"
        name="select"
        options="[object Object],[object Object],[object Object],[object Object],[object Object]"
        placeholder="Choose"
        type="select"
      >
        <option
          value="-"
        >
          Select an option
        </option>
        <option
          value="Shoes"
        >
          Shoes
        </option>
        <option
          value="Heels"
        >
          Heels
        </option>
        <option
          value="Sandals"
        >
          Sandals
        </option>
        <option
          value="Handbags"
        >
          Handbags
        </option>
      </select>
      <div
        class="sc-c25ec0ba-4 jmIHMh"
      >
        <input
          class="sc-c25ec0ba-3 ctlbxZ"
          id=":r2:"
          label="Tick if you agree with terms and conditions"
          name="checkbox"
          role="checkbox"
          type="checkbox"
        />
        <label
          class="sc-c25ec0ba-0 AOPnj"
          for=":r2:"
        >
          Tick if you agree with terms and conditions
        </label>
      </div>
      <button
        class="sc-4142dcf-1 pDeJh"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
